File: app_project/frontend/src/services/maintenance.ts
// src/services/maintenance.ts

import apiClient from "@/utils/apiClient";
import { getAuthHeaders } from "@/utils/authHeaders";
import type {
  MaintenanceRequestCreate,
  MaintenanceRequestResponse,
  MaintenanceRequestUpdate
} from "@/models/maintenance";

export const createMaintenanceRequest = async (
  requestData: MaintenanceRequestCreate
): Promise<{ success: boolean; data?: MaintenanceRequestResponse; message: string }> => {
  try {
    const response = await apiClient.post("/maintenance-requests", requestData, getAuthHeaders());
    if (response.status === 200 || response.status === 201) {
      return {
        success: true,
        data: response.data,
        message: "Maintenance request created successfully."
      };
    } else {
      return {
        success: false,
        message: "Failed to create maintenance request."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while creating the maintenance request."
    };
  }
};

export const approveMaintenanceRequest = async (
  requestId: number
): Promise<{ success: boolean; message: string }> => {
  try {
    const response = await apiClient.post(
      "/approve-maintenance-request",
      null,
      {
        ...getAuthHeaders(),
        params: { request_id: requestId }
      }
    );
    if (response.status === 200) {
      return {
        success: true,
        message: response.data.message || "Maintenance request approved successfully."
      };
    } else {
      return {
        success: false,
        message: "Failed to approve maintenance request."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while approving the maintenance request."
    };
  }
};

export const updateMaintenanceRequest = async (
  requestId: number,
  updateData: MaintenanceRequestUpdate
): Promise<{ success: boolean; data?: MaintenanceRequestResponse; message: string }> => {
  try {
    const response = await apiClient.put(
      `/update-maintenance-request/${requestId}`,
      updateData,
      getAuthHeaders()
    );
    if (response.status === 200) {
      return {
        success: true,
        data: response.data,
        message: "Maintenance request updated successfully."
      };
    } else {
      return {
        success: false,
        message: "Failed to update maintenance request."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while updating the maintenance request."
    };
  }
};

export const deleteMaintenanceRequest = async (
  requestId: number
): Promise<{ success: boolean; message: string }> => {
  try {
    const response = await apiClient.delete(
      `/delete-maintenance-request/${requestId}`,
      getAuthHeaders()
    );
    if (response.status === 200) {
      return {
        success: true,
        message: "Maintenance request deleted successfully."
      };
    } else {
      return {
        success: false,
        message: "Failed to delete maintenance request."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while deleting the maintenance request."
    };
  }
};

export const getAllMaintenanceRequests = async (): Promise<{ success: boolean; data?: MaintenanceRequestResponse[]; message: string }> => {
  try {
    const response = await apiClient.get("/maintenance-requests", getAuthHeaders());
    if (response.status === 200) {
      return {
        success: true,
        data: response.data,
        message: "Successfully retrieved maintenance requests."
      };
    } else {
      return {
        success: false,
        message: "Failed to retrieve maintenance requests."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while retrieving maintenance requests."
    };
  }
};

export const getMaintenanceRequestById = async (
  requestId: number
): Promise<{ success: boolean; data?: MaintenanceRequestResponse; message: string }> => {
  try {
    const response = await apiClient.get(`/maintenance-requests/${requestId}`, getAuthHeaders());
    if (response.status === 200) {
      return {
        success: true,
        data: response.data,
        message: "Successfully retrieved maintenance request details."
      };
    } else {
      return {
        success: false,
        message: "Failed to retrieve maintenance request details."
      };
    }
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "An error occurred while retrieving maintenance request details."
    };
  }
};

File: app_project/backend/app/api/v1/inventory/maintenance.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.core.permissions import permission_required
from app.schemas.inventory.maintenance import (
    MaintenanceRequestCreate,
    MaintenanceRequestResponse,
    MaintenanceRequestUpdate
)
from app.services.inventory.maintenance_service import (
    create_maintenance_request,
    approve_maintenance_request,
    update_maintenance_request,
    get_all_maintenance_requests,
    get_maintenance_request_by_id,
    delete_maintenance_request
)

router = APIRouter(
    prefix="/maintenance-requests",
    tags=["Maintenance Requests"],
    dependencies=[Depends(permission_required())]
)

@router.post("", response_model=MaintenanceRequestResponse)
def create_request(request_data: MaintenanceRequestCreate, db: Session = Depends(get_db)):
    """
    Create a new maintenance request.
    """
    try:
        return create_maintenance_request(db, request_data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail="Invalid maintenance request data.")

@router.get("", response_model=list[MaintenanceRequestResponse])
def list_requests(db: Session = Depends(get_db)):
    """
    Get all maintenance requests.
    """
    return get_all_maintenance_requests(db)

@router.get("/{request_id}", response_model=MaintenanceRequestResponse)
def retrieve_request(request_id: int, db: Session = Depends(get_db)):
    """
    Get a specific maintenance request by ID.
    """
    request = get_maintenance_request_by_id(db, request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Maintenance request not found.")
    return request

@router.put("/{request_id}", response_model=MaintenanceRequestResponse)
def update_request(request_id: int, update_data: MaintenanceRequestUpdate, db: Session = Depends(get_db)):
    """
    Update an existing maintenance request.
    """
    try:
        updated = update_maintenance_request(db, request_id, update_data)
        if not updated:
            raise HTTPException(status_code=404, detail="Maintenance request not found.")
        return updated
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid update data for maintenance request.")

@router.patch("/{request_id}/approve", response_model=MaintenanceRequestResponse)
def approve_request(request_id: int, db: Session = Depends(get_db)):
    """
    Approve a maintenance request.
    """
    try:
        approved = approve_maintenance_request(db, request_id)
        if not approved:
            raise HTTPException(status_code=404, detail="Maintenance request not found.")
        return approved
    except ValueError:
        raise HTTPException(status_code=400, detail="Approval failed due to invalid request status.")

@router.delete("/{request_id}")
def delete_request(request_id: int, db: Session = Depends(get_db)):
    """
    Delete a maintenance request by ID.
    """
    deleted = delete_maintenance_request(db, request_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Maintenance request not found.")
    return {"message": "Maintenance request deleted successfully."}
